### PR TITLE
[jit] Resolve with closed over variables instead of stack frame

### PR DESCRIPTION
**Pull request by driazati**
Previously we looked at the stack frame of the function that called
`script` to resolve variables. This doesn't work if someone calls script
with a function defined somewhere else that references captured
variables. We already have a mechanism to look at the closed over
variables for a function, so this changes the `rcb` to use that.

Differential Revision: [D16391346](https://our.internmc.facebook.com/intern/diff/16391346/)

### CODE REVIEW

**Review by wanchaol — APPROVED** [review]
Nice!

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -1123,12 +1121,18 @@ def script(obj, optimize=True, _frames_up=0, _rcb=None):
...
+            _rcb = _jit_internal.createResolutionCallback(_frames_up + 1)
+
         _jit_script_class_compile(qualified_name, ast, _rcb)
         _add_script_class(obj, qualified_name)
         return obj
     else:
         ast = get_jit_def(obj)
-        fn = torch._C._jit_script_compile(ast, _rcb, get_default_args(obj))
+        if _rcb is None:
+            rcb = createResolutionCallbackFromClosure(obj)
```
Comment (wanchaol) [inline torch/jit/__init__.py]: do you mean `_rcb` here instead of `rcb`?

**Comment by wanchaol** [comment]
any updates on this?

**Comment by driazati** [comment]
It breaks resolution of type comment variables since they're not actually captured by the function since they're just strings. We could fix it by combining both methods of `rcb`s, but that seems kind of hacky. I'm not sure if there's a better workaround or not

**Review by wanchaol — APPROVED** [review]
looks good.

**Pull request closed without merging**